app/src/main/java/org/wikipedia/page/leadimages/LeadImagesHandler.kt

```diff
@@ -108,35 +109,39 @@ class LeadImagesHandler(private val parentFragment: PageFragment,
-                            WikipediaApp.instance.languageState.appLanguageCodes.firstOrNull { lang -> !labelMap.containsKey(lang) }?.run {
-                                imageEditType = ImageEditType.ADD_CAPTION_TRANSLATION
-                                captionTargetPageTitle = PageTitle(imageTitle, WikiSite(Service.COMMONS_URL, this))
+                    try {
+                        lastImageTitleForCallToAction = imageTitle
+                        val isProtected = ServiceFactory.get(Constants.commonsWikiSite)
+                            .getProtectionInfoSuspend(imageTitle).query?.isEditProtected ?: false
+                        if (!isProtected) {
+                            val firstEntity = async {
+                                ServiceFactory.get(Constants.commonsWikiSite).getEntitiesByTitleSuspend(imageTitle, Constants.COMMONS_DB_NAME).first
+                            }
+                            val firstImageInfo = async {
+                                ServiceFactory.get(Constants.commonsWikiSite).getImageInfoSuspend(imageTitle, Constants.COMMONS_DB_NAME).query?.firstPage()
+                            }
+                            val labelMap = firstEntity.await()?.labels?.values?.associate { v -> v.language to v.value }.orEmpty()
+                            val depicts = ImageTagsProvider.getDepictsClaims(firstEntity.await()?.getStatements().orEmpty())
+                            imagePage = firstImageInfo.await()
+                            captionSourcePageTitle = PageTitle(imageTitle, WikiSite(Service.COMMONS_URL, it.wikiSite.languageCode))
+                            captionSourcePageTitle!!.description = labelMap[it.wikiSite.languageCode]
+                            if (!labelMap.containsKey(it.wikiSite.languageCode)) {
+                                imageEditType = ImageEditType.ADD_CAPTION
+                            }
+                            if (WikipediaApp.instance.languageState.appLanguageCodes.size >= Constants.MIN_LANGUAGES_TO_UNLOCK_TRANSLATION) {
+                                WikipediaApp.instance.languageState.appLanguageCodes.firstOrNull { lang -> !labelMap.containsKey(lang) }?.run {
+                                    imageEditType = ImageEditType.ADD_CAPTION_TRANSLATION
+                                    captionTargetPageTitle = PageTitle(imageTitle, WikiSite(Service.COMMONS_URL, this))
+                                }
+                            }
+                            if (imageEditType != ImageEditType.ADD_CAPTION && depicts.isEmpty()) {
+                                imageEditType = ImageEditType.ADD_TAGS
                             }
                         }
-                        if (imageEditType != ImageEditType.ADD_CAPTION && depicts.isEmpty()) {
-                            imageEditType = ImageEditType.ADD_TAGS
-                        }
+                        finalizeCallToAction()
+                    } catch (e: Exception) {
```

Comment:
Another reason against a big `try..catch` block is that we shouldn't catch the generic `Exception` class. When a `Job` is cancelled, it actually throws a `CancellationException`, which we _shouldn't_ catch. The CancellationException must be caught by the enclosing coroutine scope, so that it knows about the cancelled job.

Reply:
Thanks and done!